File: src/main/java/org/apache/wicket/fetch/OnChangeFetchBehavior.java
package org.apache.wicket.fetch;

import org.apache.wicket.Component;
import org.apache.wicket.markup.head.IHeaderResponse;
import org.apache.wicket.request.cycle.RequestCycle;
import org.apache.wicket.request.http.WebResponse;
import org.apache.wicket.request.resource.JavaScriptResourceReference;

import java.time.Instant;

public class OnChangeFetchBehavior extends AbstractFetchBehavior {

    // TODO all this should really be resolved from the classname
    public static final JavaScriptResourceReference RESOURCE_REFERENCE = new JavaScriptResourceReference(OnChangeFetchBehavior.class, "js/OnChangeFetchBehavior.js");

    @Override
    public void renderHead(Component component, IHeaderResponse response) {
        super.renderHead(component, response);

        response.render(new ESModuleContentHeaderItem(
                "import OnChangeFetchBehavior from \"/"+ RequestCycle.get().mapUrlFor(RESOURCE_REFERENCE, null).getPath()+"\";\n" +
                " new OnChangeFetchBehavior(\""+component.getMarkupId()+"\",\""+getCallbackUrl()+"\");"
                ,component.getMarkupId()+"-script"));
    }

    @Override
    public void onRequest() {
        // TODO look at AjaxRequestHandler, AjaxEnclosureListener and WebApplication
        RequestCycle requestCycle = RequestCycle.get();
        String encoding = getComponent().getApplication().getRequestCycleSettings().getResponseRequestEncoding();
        WebResponse response = (WebResponse) requestCycle.getResponse();
        response.disableCaching();
        response.setContentType("application/json; charset=" + encoding);
        response.write("{\"data\":\"value\",\"time\":"+ Instant.now().getEpochSecond()+"}");
        response.flush();
    }

}

File: src/main/java/org/apache/wicket/fetch/js/OnChangeFetchBehavior.js
import AbstractFetchBehavior from "./AbstractFetchBehavior.js";

export default class OnChangeFetchBehavior extends AbstractFetchBehavior{

    /**
     *
     * @param {string} id
     * @param {string} callbackUrl
     */
    constructor(id, callbackUrl) {
        super();
        this.registerCallback(this.callback);
        this.bind().then(value => {
            document.getElementById(id).addEventListener("input", function (ev) {
                /**
                 * @type {HTMLInputElement}
                 */
                const target = ev.target;
                // let data = {value: target.value};
                // TODO use json payload
                let data = new FormData();
                data.append("field2",Date.now().toString());;
                let myRequest = new Request(callbackUrl, {
                    method: 'POST',
                    body: data,
                    cache: 'no-cache',
                    headers: {
                        // 'Content-Type': 'application/json',
                        "Wicket-Fetch": true
                    }
                });
                fetch(myRequest)
                .then(response => {
                    if (!response.ok) {
                        throw new Error('HTTP error, status = ' + response.status);
                    }
                    console.log(response);
                    return response.json();
                }).then(value1 => {
                    console.log(value1);
                })

            })
        })
    }

    callback () {

    }


}